File: app/src/main/java/com/ufkoku/demo_app/ui/fragments/static_list/StaticListFragment.java
package com.ufkoku.demo_app.ui.fragments.static_list;

import android.content.Context;
import android.os.Bundle;
import android.view.View;

import com.ufkoku.demo_app.R;
import com.ufkoku.demo_app.entity.AwesomeEntity;
import com.ufkoku.demo_app.ui.common.view_state.StaticListViewState;
import com.ufkoku.demo_app.ui.lifecycle_listeners.FragmentLifecycleObserver;
import com.ufkoku.demo_app.ui.common.presenter.StaticListPresenter;
import com.ufkoku.demo_app.ui.view.DataView;
import com.ufkoku.demo_app.ui.fragments.base.IFragmentManager;
import com.ufkoku.mvp.BaseMvpFragment;
import com.ufkoku.mvp.utils.view_injection.annotation.InjectView;
import com.ufkoku.mvp.utils.view_injection.annotation.Layout;

import org.jetbrains.annotations.NotNull;

import java.util.List;

@Layout(R.layout.view_data)
public class StaticListFragment extends BaseMvpFragment<IStaticListFragment, StaticListPresenter<IStaticListFragment>, StaticListViewState> implements IStaticListFragment {

    protected static final String ARG_RETAIN = "com.ufkoku.demo_app.ui.fragments.savable.StaticListFragment.retain";

    @InjectView(R.id.view_data)
    private DataView view;

    private IStaticListFragmentWrap wrap = new IStaticListFragmentWrap(this);

    private FragmentLifecycleObserver observer = new FragmentLifecycleObserver();

    {
        subscribe(observer);
    }

    //----------------------------------------------------------------------------------------//

    @Override
    public boolean nullViews() {
        return true;
    }

    @Override
    @SuppressWarnings({"ConstantConditions"})
    public boolean retainPresenter() {
        //don't do it in real life, return constant
        Bundle args = getArguments();
        return args != null && args.getBoolean(ARG_RETAIN);
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public boolean retainViewState() {
        //don't do it in real life, return constant
        Bundle args = getArguments();
        return args != null && args.getBoolean(ARG_RETAIN);
    }

    @NotNull
    @Override
    public IStaticListFragment getMvpView() {
        return wrap;
    }

    @NotNull
    @Override
    public StaticListViewState createViewState() {
        return new StaticListViewState();
    }

    @NotNull
    @Override
    public StaticListPresenter<IStaticListFragment> createPresenter() {
        return new StaticListPresenter<>();
    }

    @Override
    public void onViewCreated(@NotNull View view, @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        this.view.setListener(new DataView.ViewListener() {
            @Override
            public void onRetainableClicked() {
                Context context = getContext();
                if (context instanceof IFragmentManager) {
                    ((IFragmentManager) context).setFragment(new Builder(true).build());
                }
            }

            @Override
            public void onSavableClicked() {
                Context context = getContext();
                if (context instanceof IFragmentManager) {
                    ((IFragmentManager) context).setFragment(new Builder(false).build());
                }
            }
        });

        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onInitialized(StaticListPresenter<IStaticListFragment> presenter, StaticListViewState viewState) {
        if (!viewState.isApplied()) {
            if (!presenter.isTaskRunning(StaticListPresenter.TASK_FETCH_DATA)) {
                presenter.fetchData();
            }
        }

        updateProgressVisibility();
    }

    //----------------------------------------------------------------------------------------//

    @Override
    public void onDataLoaded(List<AwesomeEntity> entities) {
        StaticListViewState state = getViewState();
        if (state != null) {
            state.setEntities(entities);
        }
        populateData(entities);
    }

    @Override
    public void populateData(List<AwesomeEntity> entity) {
        if (view != null) {
            view.populateData(entity);
        }
    }

    @Override
    public void setWaitViewVisible(boolean visible) {
        if (view != null) {
            view.setWaitViewVisible(visible);
        }
    }

    @Override
    public void onTaskStatusChanged(int taskId, int status) {
        updateProgressVisibility();
    }

    //---------------------------------------------------------------------------------//

    public void updateProgressVisibility() {
        StaticListPresenter<IStaticListFragment> presenter = getPresenter();
        if (presenter != null) {
            setWaitViewVisible(presenter.hasRunningTasks());
        }
    }

    //---------------------------------------------------------------------------------//

    public static class Builder {

        private boolean retainElements = false;

        public Builder() {
        }

        public Builder(boolean retainElements) {
            this.retainElements = retainElements;
        }

        public boolean isRetainElements() {
            return retainElements;
        }

        public void setRetainElements(boolean retainElements) {
            this.retainElements = retainElements;
        }

        public StaticListFragment build() {
            StaticListFragment fragment = new StaticListFragment();

            Bundle args = new Bundle();
            args.putBoolean(ARG_RETAIN, retainElements);
            fragment.setArguments(args);

            return fragment;
        }

    }

}

File: app/src/main/kotlin/com/ufkoku/demo_app/ui/fragments/static_list/StaticListFragment.kt
package com.ufkoku.demo_app.ui.fragments.static_list

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.ufkoku.demo_app.R
import com.ufkoku.demo_app.entity.AwesomeEntity
import com.ufkoku.demo_app.ui.common.presenter.StaticListPresenter
import com.ufkoku.demo_app.ui.common.view_state.StaticListViewState
import com.ufkoku.demo_app.ui.fragments.base.IFragmentManager
import com.ufkoku.demo_app.ui.lifecycle_listeners.FragmentLifecycleObserver
import com.ufkoku.demo_app.ui.view.DataView
import com.ufkoku.mvp.BaseMvpFragment
import kotlinx.android.synthetic.main.view_data.*
import java.util.*

class StaticListFragment : BaseMvpFragment<IStaticListFragmentWrap, StaticListPresenter<IStaticListFragmentWrap>, StaticListViewState>(), IStaticListFragment {

    companion object {
        protected val ARG_RETAIN = "com.ufkoku.demo_app.ui.fragments.savable.StaticListFragment.retain"
    }

    private val wrap = IStaticListFragmentWrap(this)

    private val observer = FragmentLifecycleObserver()

    init {
        subscribe(observer)
    }

    //----------------------------------------------------------------------------------------//

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.view_data, container, false)
    }

    override fun retainPresenter(): Boolean {
        //don't do it in real life, return constant
        val args = arguments
        return args != null && args.getBoolean(ARG_RETAIN)
    }

    override fun retainViewState(): Boolean {
        //don't do it in real life, return constant
        val args = arguments
        return args != null && args.getBoolean(ARG_RETAIN)
    }

    override fun getMvpView(): IStaticListFragmentWrap {
        return wrap
    }

    override fun createViewState(): StaticListViewState {
        return StaticListViewState()
    }

    override fun createPresenter(): StaticListPresenter<IStaticListFragmentWrap> {
        return StaticListPresenter()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        dataView!!.setListener(object : DataView.ViewListener {
            override fun onItemClicked(entity: AwesomeEntity) {
                presenter?.processItem(entity)
            }

            override fun onRetainableClicked() {
                val context = context
                if (context is IFragmentManager) {
                    (context as IFragmentManager).setFragment(Builder(true).build())
                }
            }

            override fun onSavableClicked() {
                val context = context
                if (context is IFragmentManager) {
                    (context as IFragmentManager).setFragment(Builder(false).build())
                }
            }
        })

        super.onViewCreated(view, savedInstanceState)
    }

    override fun onInitialized(presenter: StaticListPresenter<IStaticListFragmentWrap>, viewState: StaticListViewState) {
        //no entities in view state
        if (viewState.entities == null) {
            //load task not running
            if (!presenter.isTaskRunning(StaticListPresenter.TASK_FETCH_DATA)) {
                //load data from presenter
                presenter.fetchData()
            }
        }
        //invalidate progress visibility anyway
        updateProgressVisibility()
    }

    //----------------------------------------------------------------------------------------//

    override fun onDataLoaded(data: ArrayList<AwesomeEntity>) {
        val state = viewState
        state?.entities = data
        populateData(data)
    }

    override fun getViewWidth(): Int = dataView!!.width

    override fun onItemProcessed(result: Int) {
        Toast.makeText(context,
                       "Item processed, result is $result",
                       Toast.LENGTH_LONG).show()
    }

    override fun populateData(entities: List<AwesomeEntity>) {
        dataView?.populateData(entities)
    }

    override fun onTaskStatusChanged(taskId: Int, status: Int) {
        updateProgressVisibility()
    }

    //---------------------------------------------------------------------------------//

    fun updateProgressVisibility() {
        val presenter = presenter
        if (presenter != null) {
            dataView?.setWaitViewVisible(presenter.isTaskRunning(StaticListPresenter.TASK_FETCH_DATA))
        }
    }

    //---------------------------------------------------------------------------------//

    class Builder() {

        var isRetainElements = false

        constructor(retainElements: Boolean) : this() {
            this.isRetainElements = retainElements
        }

        fun build(): StaticListFragment {
            val fragment = StaticListFragment()

            val args = Bundle()
            args.putBoolean(ARG_RETAIN, isRetainElements)
            fragment.arguments = args

            return fragment
        }

    }

}